peps/pep-0797.rst

```diff
@@ -0,0 +1,672 @@
+memory in an interpreter different from the one that created it.
+
+Immortality Has No Impact on Reference Counting Performance
+-----------------------------------------------------------
+
+If the primary issue with cross-interpreter reference counting is thread-safety,
+why not make reference counting atomic? Unfortunately, atomic operations are
+slower than their non-atomic counterparts, so this would hurt overall
+performance. :pep:`703` managed to solve this issue using biased reference
+counting, splitting the reference count field into two, but at the cost of
+breaking the stable C API. Atomic reference counting on its own also doesn't
+solve the per-interpreter deallocator issue as mentioned previously.
+
+Immortality is a Powerful Tool to Wield
+***************************************
+
+In general, immortal objects seem to have a lot of applications that aren't
+immediately clear to CPython maintainers. For example, `Nanobind used immortal objects`_
+in its original :term:`free threading` implementation. So, it seems like immortality
+has interesting use-cases in third-party libraries.
+
+.. _Nanobind used immortal objects: https://github.com/wjakob/nanobind/pull/695
+
+Specification
+=============
+
+Changes to the Internal API
+***************************
+
+Internally, there are two functions for making an object immortal:
+``_Py_SetImmortal`` and ``_Py_SetImmortalUntracked``.
+Because the interpreter might want to do its own things with immortal
+objects, either now or in the future, the interpreter state should yield
+control of a user-defined immortal object to the internal API. So,
+the existing internal functions will gain an extra case to remove an object
+from the user-defined immortals, so the interpreter won't try and manually
+free it later. A new internal function, ``_Py_SetImmortalKnown``, will be
+used instead for places where it's certain that objects could not have
+been made immortal by a user.
+
+Changes to String Interning
+***************************
+
+Currently, the string interning implementation is reliant
+on the idea that it controls when a string will become immortal.
+Since a user can now immortalize a string on their own, string
+interning has to be made "self-aware," meaning that it has to be
+able to dynamically figure out if its immortal or not.
+
+The only place where this needs to change is in
+:c:func:`PyUnicode_CHECK_INTERNED`. If the current state is
+``SSTATE_INTERNED_MORTAL``, but the object is immortal, then it should
+return ``SSTATE_INTERNED_IMMORTAL`` instead.
+
+Changes to Object Finalization
+******************************
+
+The C API function, :c:func:`PyObject_CallFinalizer`, must be able to
+de-duplicate calls to immortal objects, even if they're non-GC types.
+
+Immortalization APIs
+********************
+
+This PEP introduces two new semi-public APIs for making an object immortal:
+:c:func:`PyUnstable_Immortalize` and :func:`sys._immortalize`.
```

Comment:
They're not semi-public, one is private and the other is unstable. We should have clarity over our internal API contracts.
```suggestion
This PEP introduces two new APIs for making an object immortal:
:c:func:`PyUnstable_Immortalize` and :func:`sys._immortalize`.
```